File: backend/app/models/chapter.py
"""Chapter model."""
import uuid
from datetime import datetime
from sqlalchemy import String, Integer, Text, DateTime, ForeignKey, func
from sqlalchemy.orm import Mapped, mapped_column
from app.database import Base


class Chapter(Base):
    __tablename__ = "chapters"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    project_id: Mapped[str] = mapped_column(String(36), ForeignKey("projects.id", ondelete="CASCADE"), index=True)
    outline_id: Mapped[str | None] = mapped_column(String(36), ForeignKey("outlines.id", ondelete="SET NULL"), nullable=True)
    title: Mapped[str] = mapped_column(String(500), default="")
    content: Mapped[str | None] = mapped_column(Text, nullable=True)
    word_count: Mapped[int] = mapped_column(Integer, default=0)
    chapter_index: Mapped[int] = mapped_column(Integer, default=0)
    status: Mapped[str] = mapped_column(String(50), default="draft")  # draft/polished/final
    writing_style_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    model_override: Mapped[str | None] = mapped_column(String(100), nullable=True)
    skill_override: Mapped[str | None] = mapped_column(String(100), nullable=True)
    narrative_perspective_override: Mapped[str | None] = mapped_column(String(50), nullable=True)
    expansion_plan: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON
    continuation_mode: Mapped[str] = mapped_column(String(50), default="auto")  # auto/new/continue
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())

File: frontend/src/types/chapter.ts
export interface ChapterSummary {
  id: string; index: number; title: string; word_count: number; status: string;
}
export interface Chapter {
  id: string; project_id: string; chapter_index: number; title: string;
  content: string | null; word_count: number; status: string;
  outline_id?: string | null; writing_style_id?: string | null;
  created_at?: string; updated_at?: string;
}
export interface ChapterGenerateRequest {
  project_id: string; current_chapter_index?: number; title?: string;
  genre?: string; description?: string; outlines?: Record<string, unknown>[];
  characters?: Record<string, unknown>[]; chapters?: Record<string, unknown>[];
  world_setting?: Record<string, unknown>; foreshadows?: Record<string, unknown>[];
  plot_memory?: Record<string, unknown>[]; chapter_analyses?: Record<string, unknown>[];
  generation_config?: Record<string, unknown>; writing_style_id?: string | null;
  active_skill?: string | null; human_feedback?: string | null;
}
